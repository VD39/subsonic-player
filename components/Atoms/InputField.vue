<script setup lang="ts">
withDefaults(
  defineProps<{
    error?: string;
    hideLabel?: boolean;
    id: string;
    label: string;
    modelValue: string | string[];
    placeholder?: string;
    required?: boolean;
    type?: string;
  }>(),
  {
    error: undefined,
    placeholder: undefined,
    type: 'text',
  },
);

const emit = defineEmits(['update:modelValue']);

function onInputInput(event: Event) {
  emit('update:modelValue', (event.target as HTMLInputElement).value);
}
</script>

<template>
  <div
    :class="[
      'column',
      $style.inputField,
      {
        [$style.error]: error,
      },
    ]"
  >
    <label
      ref="label"
      :class="[
        'mBXS',
        'sentenceCase',
        'smallFont',
        $style.label,
        {
          visuallyHidden: hideLabel,
        },
      ]"
      :for="id"
    >
      {{ label }}
      <span v-if="required" ref="required" :class="$style.required">*</span>
    </label>

    <input
      :id="id"
      ref="input"
      :type="type"
      autocomplete="off"
      :value="modelValue"
      :required="required"
      :placeholder="placeholder"
      :class="$style.input"
      @input="onInputInput"
    />

    <p v-if="error" ref="error" class="smallFont sentenceCase">
      {{ error }}
    </p>
  </div>
</template>

<style module>
.inputField {
  --label-color: var(--secondary-font-color);
  --input-border-color: var(--border-color);
  --input-width: var(--width-height-100);

  width: var(--input-width);
  color: var(--label-color);
}

.error {
  --label-color: var(--error-color);
  --input-border-color: var(--error-color);
}

.label {
  display: inline-block;
  color: var(--secondary-font-color);
}

.input {
  position: relative;
  width: var(--input-width);
  padding: calc(var(--space-12) - 2px) var(--space-12);
  color: var(--body-font-color);
  background-color: var(--secondary-background-color);
  border: 1px solid var(--input-border-color);
  border-radius: var(--border-radius-small);

  &::placeholder {
    opacity: 0.8;
  }
}

.required {
  color: var(--error-color);
}
</style>
