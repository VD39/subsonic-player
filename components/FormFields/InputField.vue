<script setup lang="ts">
withDefaults(
  defineProps<{
    errorMessage?: string;
    hasError?: boolean;
    hideLabel?: boolean;
    id: string;
    label: string;
    modelValue: string | string[];
    placeholder?: string;
    required?: boolean;
    type?: string;
  }>(),
  {
    errorMessage: undefined,
    placeholder: undefined,
    type: 'text',
  },
);

const emit = defineEmits(['update:modelValue']);

function updateModel(event: Event) {
  emit('update:modelValue', (event.target as HTMLInputElement).value);
}
</script>
<template>
  <div
    :class="[
      $style.inputField,
      {
        [$style.error]: hasError,
      },
    ]"
  >
    <label
      ref="label"
      :class="[
        $style.label,
        {
          'visually-hidden': hideLabel,
        },
      ]"
      :for="id"
    >
      {{ label }}
      <span v-if="required" ref="required" :class="$style.required">*</span>
    </label>

    <input
      :id="id"
      ref="input"
      :type="type"
      autocomplete="off"
      :value="modelValue"
      :required="required"
      :placeholder="placeholder"
      :class="$style.input"
      @input="updateModel"
    />

    <div v-if="hasError && errorMessage" ref="errorMessage">
      <p :class="$style.errorMessage">
        {{ errorMessage }}
      </p>
    </div>
  </div>
</template>

<style module>
.inputField {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.label {
  display: inline-block;
  color: var(--secondary-font-color);

  &::first-letter {
    text-transform: capitalize;
  }
}

.input {
  position: relative;
  width: 100%;
  padding: calc(var(--space-12) - 1px) var(--space-12);
  color: var(--body-font-color);
  background-color: var(--secondary-background-color);
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-s);

  .error & {
    border-color: var(--error-color);

    &:focus {
      border-color: var(--error-color);
    }
  }

  &::placeholder {
    opacity: 0.5;
  }
}

.required {
  color: var(--error-color);
}

.errorMessage {
  margin-top: var(--space-2);
  font-size: var(--small-font-size);
  color: var(--error-color);

  &::first-letter {
    text-transform: capitalize;
  }
}
</style>
