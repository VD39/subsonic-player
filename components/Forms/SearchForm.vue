<script setup lang="ts">
import InputField from '@/components/FormFields/InputField.vue';
import IconButton from '@/components/Buttons/IconButton.vue';

const emit = defineEmits(['submit']);

const formInputs = {
  query: {
    validationRules: {
      required: true,
    },
  },
};

const form = createForm(formInputs);

async function submitForm() {
  validateInputs(form);

  if (!form.isValid.value) {
    return;
  }

  const query = replaceSpacesWithCharacter(
    form.fields.query.value.value as string,
  ).toLowerCase();

  form.fields.query.value.value = '';

  emit('submit', query);
}
</script>

<template>
  <form novalidate @submit.prevent="submitForm">
    <div :class="$style.inner">
      <InputField
        :id="form.fields.query.id"
        v-model="form.fields.query.value.value"
        :class="$style.input"
        :label="form.fields.query.label"
        placeholder="Enter query"
        :required="form.fields.query.required"
        hide-label
      />

      <IconButton type="submit" :class="$style.button" icon="PhMagnifyingGlass">
        Login
      </IconButton>
    </div>
  </form>
</template>

<style module>
.inner {
  @mixin align-center;

  position: relative;
}

.input {
  input {
    padding-right: var(--space-40);
  }
}

.button {
  margin-left: calc(var(--space-40) * -1);
}
</style>
