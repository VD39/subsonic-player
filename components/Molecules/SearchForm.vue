<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import InputField from '@/components/Atoms/InputField.vue';

const emit = defineEmits<{
  submit: [value: string];
}>();

const formInputs = {
  query: {
    validationRules: {
      required: true,
    },
  },
};

const form = createForm(formInputs);

async function onFormSubmit() {
  validateInputs(form);

  if (!form.isValid.value) {
    return;
  }

  const query = replaceSpacesWithCharacter(
    form.fields.query.value.value as string,
  ).toLowerCase();

  form.fields.query.value.value = '';

  emit('submit', query);
}
</script>

<template>
  <form novalidate @submit.prevent="onFormSubmit">
    <div class="centerItems">
      <InputField
        :id="form.fields.query.id"
        v-model="form.fields.query.value.value"
        :class="$style.inputField"
        hide-label
        :label="form.fields.query.label"
        placeholder="Enter query"
        :required="form.fields.query.required"
      />

      <ButtonLink :class="$style.buttonLink" :icon="ICONS.search" type="submit">
        Login
      </ButtonLink>
    </div>
  </form>
</template>

<style module>
.inputField {
  input {
    padding-right: calc(var(--space-24) * 2);
  }
}

.buttonLink {
  margin-left: calc(var(--space-24) * -2);
}
</style>
