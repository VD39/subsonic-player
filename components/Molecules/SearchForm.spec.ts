import type { VueWrapper } from '@vue/test-utils';

import { mount } from '@vue/test-utils';

import InputField from '@/components/Atoms/InputField.vue';

import SearchForm from './SearchForm.vue';

function factory(props = {}) {
  return mount(SearchForm, {
    props: {
      ...props,
    },
  });
}

describe('SearchForm', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when form is invalid', () => {
    beforeEach(async () => {
      await wrapper.trigger('submit');
    });

    it('does not emit submit event', () => {
      expect(wrapper.emitted('submit')).toBe(undefined);
    });
  });

  describe('when form is valid', () => {
    beforeEach(async () => {
      wrapper.findComponent(InputField).vm.$emit('update:modelValue', 'query');
      await wrapper.trigger('submit');
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('emits submit event with the form values', () => {
      expect(wrapper.emitted('submit')).toEqual([['query']]);
    });
  });
});
