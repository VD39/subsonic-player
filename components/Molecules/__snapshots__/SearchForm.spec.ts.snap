// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="centerItems">
    <div class="column inputField inputField"><label class="mBXS sentenceCase smallFont label visuallyHidden" for="query">query <span class="required">*</span></label><input id="query" type="text" autocomplete="off" required="" placeholder="Enter query" class="input" value="">
      <!--v-if-->
    </div><button class="centerItems buttonLink buttonLink buttonLink" type="submit">
      <phmagnifyingglass class="icon" aria-hidden="true" size="24" weight="regular"></phmagnifyingglass><span class="clamp text visuallyHidden"> Login </span>
    </button>
  </div>
</form>"
`;

exports[`SearchForm > when form is valid > matches the snapshot 1`] = `
"<form novalidate="">
  <div class="centerItems">
    <div class="column inputField inputField"><label class="mBXS sentenceCase smallFont label visuallyHidden" for="query">query <span class="required">*</span></label><input id="query" type="text" autocomplete="off" required="" placeholder="Enter query" class="input" value="">
      <!--v-if-->
    </div><button class="centerItems buttonLink buttonLink buttonLink" type="submit">
      <phmagnifyingglass class="icon" aria-hidden="true" size="24" weight="regular"></phmagnifyingglass><span class="clamp text visuallyHidden"> Login </span>
    </button>
  </div>
</form>"
`;
